crypto-sodium/lib/Crypto/Hash/Internal.hs

```diff
@@ -0,0 +1,60 @@
+-- SPDX-FileCopyrightText: 2021 Serokell
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+{-# LANGUAGE ExplicitNamespaces, TypeOperators, TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+-- | Internals of @crypto_generichash@.
+module Crypto.Hash.Internal
+  ( HashBlake2b
+  , blake2b
+  ) where
+
+import Prelude hiding (length)
+
+import Data.ByteArray (ByteArray, ByteArrayAccess, length, withByteArray)
+import Data.ByteArray.Sized (SizedByteArray, allocRet)
+import Data.Proxy (Proxy (Proxy))
+import Foreign.Ptr (nullPtr)
+import GHC.TypeNats (KnownNat, natVal, type (<=))
+
+import qualified Libsodium as Na
+
+-- | Hash returned by 'blake2b'.
+--
+-- This type is parametrised by hash size in bytes and the actual data type
+-- that contains bytes. This can be, for example, a @ByteString@.
+--
+-- Length must be between 16 and 64 bytes.
+type HashBlake2b len a = SizedByteArray len a
+
+-- | Hash a message using BLAKE2b.
+blake2b
+  ::  forall pt key hashBytes len.
+      ( ByteArrayAccess pt
+      , ByteArrayAccess key
+      , ByteArray hashBytes
+      , KnownNat len
+      , Na.CRYPTO_GENERICHASH_BYTES_MIN <= len
+      , len <= Na.CRYPTO_GENERICHASH_BYTES_MAX
+      )
+  => pt  -- ^ Message to hash
+  -> Maybe key -- ^ Hash key
+  -> IO (Maybe (HashBlake2b len hashBytes))
```

Comment:
Please, remove the `Maybe`. This function [always returns 0](https://github.com/jedisct1/libsodium/blob/5df53ce62b41ea3c624f0565fe5f088aa295801e/src/libsodium/crypto_generichash/blake2b/ref/blake2b-ref.c#L368).

I know libsodium docs don’t say this explicitly (unlike NaCl docs), but, generally, their docs are of a rather low quality, and when they don’t document the return value it usually means that it is always 0. So we here have to rely on the implementation occasionally to provide a sensible user interface. Just leave a comment in the code (again, see how it is done for `NaCl.Hash`).

Reply:
See, I'm not that familiar with libsodium, so I wasn't sure about that. Reasonably speaking, it shouldn't fail (at least if hash length is between min and max inclusive), but yeah, docs weren't very specific about the return value. Done.